PARQUET-1630: Resolve Bloom filter spec concerns (#146)
apache/parquet-format: BloomFilter.md

parameters to end users.

#### File Format

> Each multi-block Bloom filter is required to work for only one column chunk. The data of a multi-block Bloom

Reviewer: Nit: Paragraphs should be separated from previous content by a blank line.

Author: Thanks, will update in next commit.

Reviewer: Looks like you missed some of the missing blank lines.

> filter contains a header of Bloom filter, which must includes the size of the filter in bytes, the algorithm,

Reviewer 2: when you say "a header of Bloom filter", do you mean "the header of one Bloom filter"?

Also, "include the size", not "includes the size".

> the hash function, and the Bloom filter bitset. The offset in column chunk metadata points to the start of

Reviewer 2: ```suggestion
the hash function, and the Bloom filter bitset. The offset in column chunk metadata points to the start of the
```

the Bloom filter header. 

> Here are Bloom filter definitions in thrift:

Reviewer 2: ```suggestion
Here are the Bloom filter definitions in thrift:
```

parameters to end users.

#### File Format

Each multi-block Bloom filter is required to work for only one column chunk. The data of a multi-block Bloom
filter contains the header of one Bloom filter, which must include the size of the filter in bytes, the algorithm,
the hash function, the compression and the Bloom filter bitset. The offset in column chunk metadata points to the start of

> the Bloom filter header. 

Reviewer: This wording is strange because the header doesn't include the bloom filter bitset. I think you should change this to "The data of a multi-block bloom filter consists of the bloom filter header followed by the bloom filter bitset. The bloom filter header encodes the size of the bloom filter bit set in bytes that is used to read the bitset."

No need to call out the other header fields.

Author: Sound good. Updated.

Author: @rdblue, ready for review now.

```

> The Bloom filter data is stored right after pages indexes, the file layout is look like:

Reviewer 2: ```suggestion
The Bloom filter data is stored right after the page indexes, and the file layout looks like:
```

Reviewer 2: This is still incorrect. Can you look at my suggestion more carefully and identify what you missed?

Author: Oops.

Reviewer 3: @jbapple  @chenjunjiedada  Hi for some reason this statement has now says "before the page indexes" but the images has the bloom filter after the pages indices. Should be this be updated?

> The Bloom filter data is stored right after the page indexes, and the file layout looks like:

Reviewer: Why is this stored after page indexes? The page index data is going to be used more often and is smaller, so I think it should be just before the header and bloom filters should be before the indexes.

Why is this layout required? Was there performance testing done adding the bloom filters here instead of at the start of row groups?

Author: There is no specific reason why the bloom filter is stored after page indexes. IIUC, both page indexes and bloom filter are random reads, will they impact each other? 

Regarding store bloom filter at the footer, that is because when using HDFS seeking in a parquet file may require remote reads which consuming remote I/O and bandwidth and we can't reading parquet footer in any case. So that we can save some  I/O when storing bloom filters in the footer.

Reviewer: Page indexes aren't necessarily going to require random reads. Some readers guess the length of the footer and read, for example, 1MB of data at the end of the file to avoid a backward seek. If that includes the entire footer and page indexes, then it doesn't require a second backward seek. And it will become more likely that the indexes are in that 1MB chunk when we update Parquet to use truncated lower and upper bounds.

Anyway, the point is that the bloom filters shouldn't be placed between the page indexes and the footer.

> because when using HDFS seeking in a parquet file may require remote reads which consuming remote I/O and bandwidth

I don't think this is true. Usually, the footer is read remotely.

A task is created for an HDFS block and assigned where that block is local. When that task runs, it will read the footer, which usually probably remote, find which row groups are in the assigned block, and then read the matching row groups that are probably local. Since bloom filter data is large, I think you would end up with more remote reads in HDFS by locating them near the footer.

Keeping bloom filters in memory to be placed near the footer is also more expensive for memory consumption because Parquet can't just write out the filter data and reuse that memory.

I'm not sure what the right choice is here.

@jbapple, @julienledem, and @zivanfi: any opinions about where bloom filters should be located?

Author: Understood your point, thanks for the elaboration.

Reviewer 2: @rdblue Thank you for asking. I'm not yet well informed enough about the details to have an opinion.

Author: Just updated the layout according to Ryan's comments, I can update the layout if there is any update after discussion.

Reviewer: @chenjunjiedada, to get this done, let's allow both as an option. So bloom filters may be located either between row groups (e.g., bloom filter col 1, bloom filter col 3, column chunk 1, column chunk 2, column chunk 3) OR may be located before the page indexes after all row groups.

In both cases, bloom filters should be grouped by row group with data for each column in the same order as the file schema. That will reduce seeks for row group data when processing a subset of the row groups, which is the common access pattern from processing splits.

Author: Agreed, maybe we can provide an option in writer to select the location.

Author: @rdblue, format updated. Please take a look.
